src/blake.js

```diff
@@ -1,56 +1,79 @@
-'use strict'
+// @flow
+import * as blake from 'blakejs'
+import type { Hash, HashUpdate, HashTable, HashBuilder } from './types'
+import type { Code } from 'multihashes/lib/constants'
 
-const blake = require('blakejs')
-const minB = 0xb201
-const minS = 0xb241
+const minB: Code = 0xb201
+const minS: Code = 0xb241
 
-var blake2b = {
+type BlakeCtx = {
+  b: Uint8Array,
+  h: Uint32Array,
+  t: number,
+  c: number,
+  outlen: number
+}
+type Blake2Hash = Buffer
+type BlakeHasher = {
+  init(size: number, key: ?number): BlakeCtx,
+  update(ctx: BlakeCtx, input: Uint8Array): void,
+  digest(ctx: BlakeCtx): Uint8Array
+}
+
+const blake2b: BlakeHasher = {
   init: blake.blake2bInit,
   update: blake.blake2bUpdate,
   digest: blake.blake2bFinal
 }
 
-var blake2s = {
+const blake2s: BlakeHasher = {
   init: blake.blake2sInit,
   update: blake.blake2sUpdate,
   digest: blake.blake2sFinal
 }
 
-class B2Hash {
+class B2Hash implements Hash {
+  ctx: BlakeCtx | null
+  hf: BlakeHasher
+
   constructor (size, hashFunc) {
     this.hf = hashFunc
     this.ctx = this.hf.init(size, null)
   }
 
-  update (buf) {
+  static new (size, hashFunc): HashUpdate {
+    return new B2Hash(size, hashFunc)
+  }
+
+  update (buf: Buffer): Hash {
     if (this.ctx === null) {
       throw new Error('blake2 context is null. (already called digest?)')
     }
     this.hf.update(this.ctx, buf)
     return this
   }
 
-  digest () {
+  digest (): Blake2Hash {
     const ctx = this.ctx
     this.ctx = null
+    if (ctx === null) {
+      throw Error('blake2 context is null. (already called digest?)')
+    }
     return Buffer.from(this.hf.digest(ctx))
   }
 }
 
-function addFuncs (table) {
-  function mkFunc (size, hashFunc) {
-    return () => new B2Hash(size, hashFunc)
+export const addFuncs = (table: HashTable) => {
+  const mkFunc = (size: number, hashFunc: BlakeHasher): HashBuilder => {
+    return (): HashUpdate => B2Hash.new(size, hashFunc)
   }
 
-  var i
+  // I don't like using any here but the only way I could get the types to work here.
```

Comment:
More out of curiosity - what was the problem? Shouldn't `Code` be just an alias for `number`?

Reply:
They were opaque type aliases which behave like nominal types. In that module context any number can be treated as Code but outside of that context Code is a subtype of number and only way to get hold of value of that tipe is by getting it from that module either from exposed constant annotated as Code or via function that returns Code. This gives you a way to say have `parseCode : number -> ?Code` function that can return passed value back but ensuring that it’s a valid Code. Kind of typically you’d have `isCode` function but such that type checker then enforce typings